Replace _Sequence by _FakeSequence

p-snft opened the pull request:
The auto-stretch feature of the _Sequence contained hidden magic, as it tracked the last "entry" that has been accessed and pretended to have a length based on this. So, outcomes depend on something that is assumed to be a read access. Also, self.data was erived from UserList but never really supported all of the expected functionality in a consistent way.

The new _FakeSequence is a lot simpler. In particular it has an explicit (fake) length that needs to be set before casting to (stretched) lists. The default length is not set, because member functions using the same names as in `numpy.array` can help to work without an iterator.

Fixes #1100.

pep8speaks commented:
Hello @p-snft! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:















* In the file [`tests/test_plumbing.py`](https://github.com/oemof/oemof-solph/blob/3e94cccf4cfb38233feaac6055306f8ca1432162/tests/test_plumbing.py):

> [Line 17:22](https://github.com/oemof/oemof-solph/blob/3e94cccf4cfb38233feaac6055306f8ca1432162/tests/test_plumbing.py#L17): [E711](https://duckduckgo.com/?q=pep8%20E711) comparison to None should be 'if cond is None:'
> [Line 20:22](https://github.com/oemof/oemof-solph/blob/3e94cccf4cfb38233feaac6055306f8ca1432162/tests/test_plumbing.py#L20): [E711](https://duckduckgo.com/?q=pep8%20E711) comparison to None should be 'if cond is None:'
> [Line 28:23](https://github.com/oemof/oemof-solph/blob/3e94cccf4cfb38233feaac6055306f8ca1432162/tests/test_plumbing.py#L28): [E711](https://duckduckgo.com/?q=pep8%20E711) comparison to None should be 'if cond is None:'

uvchik commented:
> TODO: solph.processing still expects the auto-stretching to have happened.

Where is auto-stretching needed and why are all tests fine? I would expect a failing test that will be fine if we have fixed that issue.

p-snft commented:
> > TODO: solph.processing still expects the auto-stretching to have happened.
> 
> Where is auto-stretching needed and why are all tests fine? I would expect a failing test that will be fine if we have fixed that issue.

I fixed that already (in [3e94ccc](https://github.com/oemof/oemof-solph/pull/1101/commits/3e94cccf4cfb38233feaac6055306f8ca1432162)).

uvchik reviewed (approved):
Looks fine to me and in my opinion things are clearer now.
